src/registry/artifactory.rs

```diff
@@ -26,191 +27,166 @@ pub struct Artifactory {
-            response.status() != 401,
-            "Unauthorized - please provide registry credentials with `buffrs login`",
-        );
+impl TryFrom<Response> for ValidatedResponse {
+    type Error = miette::Report;
 
+    fn try_from(value: Response) -> Result<Self, Self::Error> {
         ensure!(
-            response.status().is_success(),
-            "Failed to fetch {dependency}: {}",
-            response.status()
+            !value.status().is_redirection(),
+            "remote server attempted to redirect request - is this registry URL valid?"
         );
 
-        let headers = response.headers();
-        let content_type = headers
-            .get(&reqwest::header::CONTENT_TYPE)
-            .wrap_err("missing header in response")?;
-
         ensure!(
-            content_type == reqwest::header::HeaderValue::from_static("application/x-gzip"),
-            "Server response has incorrect mime type: {content_type:?}"
+            value.status() != 401,
+            "unauthorized - please provide registry credentials with `buffrs login`"
         );
 
-        tracing::debug!("downloaded dependency {dependency}");
-
-        let tgz = response.bytes().await.wrap_err("Failed to download tar")?;
-
-        Package::try_from(tgz).wrap_err_with(|| {
-            format!(
-                "Failed to process dependency {}@{}",
-                dependency.package, version
-            )
-        })
+        value.error_for_status().into_diagnostic().map(Self)
     }
+}
 
-    /// Publishes a package to artifactory
-    pub async fn publish(&self, package: Package, repository: String) -> eyre::Result<()> {
-        let artifact_uri: Url = format!(
-            "{}/{}/{}/{}-{}.tgz",
-            self.registry,
-            repository,
-            package.name(),
-            package.name(),
-            package.version(),
-        )
-        .parse()
-        .wrap_err("Failed to construct artifact uri")?;
+fn resolve_url(dependency: Dependency) -> miette::Result<Url> {
+    let version = super::dependency_version_string(&dependency)?;
 
-        let mut request = self.client.put(artifact_uri).body(package.tgz.clone());
+    let path = dependency.manifest.registry.path().to_owned();
 
-        if let Some(token) = &self.token {
-            request = request.bearer_auth(token);
-        }
+    let mut url = dependency.manifest.registry.clone();
+    url.set_path(&format!(
+        "{}/{}/{}/{}-{}.tgz",
+        path, dependency.manifest.repository, dependency.package, dependency.package, version
+    ));
 
-        let response = request
-            .send()
-            .await
-            .wrap_err("Failed to upload release to artifactory")?;
+    Ok(url.into())
+}
 
-        ensure!(
-            !response.status().is_redirection(),
-            "Remote server attempted to redirect publish request - is the registry URL valid?"
-        );
+async fn convert_to_package(response: ValidatedResponse) -> miette::Result<Package> {
```

Comment:
i dislike the fundamental changes to the artifactory impl, this looks way more confusing to me as it did before

Reply:
This was a poor man's attempt at minimising the risk someone would make a package out of a non-validated response, by encapsulating that logic into a standalone function, but can revert this back to inline code.